rktlet/runtime/exec.go

```diff
@@ -17,25 +17,175 @@ limitations under the License.
 package runtime
 
 import (
+	"bytes"
 	"errors"
+	"io"
+	"io/ioutil"
+	"os/exec"
+	"syscall"
+
+	"github.com/golang/glog"
+	"github.com/kr/pty"
+	"github.com/kubernetes-incubator/rktlet/rktlet/cli"
+	utilexec "k8s.io/kubernetes/pkg/util/exec"
 
 	"golang.org/x/net/context"
 
-	runtimeApi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+	"k8s.io/kubernetes/pkg/kubelet/server/streaming"
+	"k8s.io/kubernetes/pkg/util/term"
 )
 
-func (r *RktRuntime) Exec(ctx context.Context, req *runtimeApi.ExecRequest) (*runtimeApi.ExecResponse, error) {
-	return nil, errors.New("TODO")
+func (r *RktRuntime) Attach(ctx context.Context, req *runtimeapi.AttachRequest) (*runtimeapi.AttachResponse, error) {
+	// TODO, the second parameter here needs to be retrieved from the
+	// `ContainerConfig` associated with the req.ContainerID
+	return r.streamServer.GetAttach(req, true)
+}
+
+func (r *RktRuntime) Exec(ctx context.Context, req *runtimeapi.ExecRequest) (*runtimeapi.ExecResponse, error) {
+	return r.streamServer.GetExec(req)
+}
+
+type nopWriteCloser bytes.Buffer
+
+func (n nopWriteCloser) Bytes() []byte {
+	return n.Bytes()
+}
+
+func (n nopWriteCloser) Write(p []byte) (int, error) {
+	return n.Write(p)
+}
+
+func (nopWriteCloser) Close() error {
+	return nil
+}
+
+func (r *RktRuntime) ExecSync(ctx context.Context, req *runtimeapi.ExecSyncRequest) (*runtimeapi.ExecSyncResponse, error) {
+	var stdout, stderr nopWriteCloser
+	nopStdin := ioutil.NopCloser(bytes.NewReader([]byte{}))
+
+	err := r.execShim.Exec(req.GetContainerId(), req.GetCmd(), nopStdin, stdout, stderr, false, make(chan term.Size))
```

Comment:
Tihs is `ExecSync` which is much different from `Exec`. `ExecSync` never has a tty, and never has a streaming stdin (only a `Cmd`). That's why there's nop-readers and no tty, the `ExecSyncRequest` doesn't have them in any case.